script/testing/micro_bench/micro_benchmarks_runner.py

```diff
@@ -0,0 +1,165 @@
+
+        ret_val, err = self._execute_benchmark(cmd)
+
+        if ret_val == 0:
+            convert_result_xml(bench_name, output_file)
+        else:
+            LOG.error("Unexpected failure of {BENCHMARK} [ret_val={RET_CODE}]".format(
+                BENCHMARK=bench_name, RET_CODE=ret_val))
+            LOG.error(err)
+
+        # return the process exit code
+        return ret_val
+
+    def _build_benchmark_cmd(self, bench_name, output_file, enable_perf):
+        """ Given the arguments passed in this will construct the command
+        necessary to execute the microbenchmark test """
+        benchmark_path = os.path.join(self.config.benchmark_path, bench_name)
+        cmd = "{BENCHMARK} --benchmark_min_time={MIN_TIME} " + \
+              " --benchmark_format=json" + \
+              " --benchmark_out={OUTPUT_FILE}"
+        cmd = cmd.format(BENCHMARK=benchmark_path,
+                         MIN_TIME=self.config.min_time, OUTPUT_FILE=output_file)
+
+        if enable_perf:
+            if not is_package_installed('perf'):
+                raise Exception("Missing perf binary. Please install package.")
+            perf_cmd = generate_perf_command(bench_name)
+            cmd = "{PERF_CMD} {CMD}".format(PERF_CMD=perf_cmd, CMD=cmd)
+
+        if not is_package_installed('numactl', '--show') and not self.config.is_local:
+            raise Exception("Missing numactl binary. Please install package")
+            # pass # if you are running locally but want to access Jenkins artifacts comment the above line
+        numa_cmd = generate_numa_command()
+        cmd = "{NUMA_CMD} {CMD}".format(NUMA_CMD=numa_cmd, CMD=cmd)
+        return cmd
+
+    def _execute_benchmark(self, cmd):
+        """ Execute the microbenchmark command provided """
+        LOG.debug("Executing command [num_threads={}]: {}".format(self.config.num_threads, cmd))
+        try:
+            output = subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True)
+            pretty_format_json = json.dumps(json.loads(output.decode('utf8').replace("'", '"')), indent=4)
+            LOG.debug("OUTPUT: {}".format(pretty_format_json))
+            return 0, None
+        except subprocess.CalledProcessError as err:
+            return err.returncode, err
+        except Exception as err:
+            return 1, err
+
+    def create_local_dirs(self):
+        """ 
+        This will create a directory for the build in the LOCAL_REPO_DIR. 
+        Each time the microbenchmark script is run it will create another dir
+        by incrementing the last dir name created. If the script is run 3 times
+        the LOCAL_REPO_DIR will have directories named 001 002 003 each
+        containing the json Google benchmark result file.
+        """
+        build_dirs = next(os.walk(LOCAL_REPO_DIR))[1]
+        last_build = max(build_dirs) if build_dirs else '000'
+        next_build = os.path.join(LOCAL_REPO_DIR, "{:03}".format(int(last_build) + 1))
```

Comment:
Not sure, but does Jenkins has this limitation of the up-to-999 build number stuff? What if it goes to 4 digits onwards?

Reply:
This is for local execution only. So the user would have to run it locally 1001 times and the fix would be to delete one of the runs so I'm not too worried.